presto-native-execution/presto_cpp/main/LinuxMemoryChecker.cpp

```diff
@@ -0,0 +1,128 @@
+// However we found out that usage_in_bytes is a fuzz value
+// and has a chance for the sync to occur after the shrink
+// polling interval. This would result in double shrinks.
+//
+// Therefore we decided on values from the memory.stat file
+// that are real time statistics. At first we tried to use
+// the calculation suggested by the kernel team RSS+CACHE(+SWAP)
+// However we noticed that this value was not closely related to the
+// value in usage_in_bytes which is used to OOMKill. We then looked
+// at all of the values in the stat file and decided that
+// inactive_anon + active_anon moves closest to that of
+// usage_in_bytes
+//
+// NOTE: We do not know if cgroup V2 memory.current is a fuzz
+// value. It may be better than what we currently use. For
+// consistency we will match cgroup V1 and change if
+// necessary.
+
+int64_t LinuxMemoryChecker::systemUsedMemoryBytes() {
+#ifdef __linux__
+  static std::string statFile;
+  size_t memAvailable = 0;
+  size_t memTotal = 0;
+  size_t inactiveAnon = 0;
+  size_t activeAnon = 0;
+  boost::cmatch match;
+  std::array<char, 50> buf;
+  static const boost::regex inactiveAnonRegex(R"!(inactive_anon\s*(\d+)\s*)!");
+  static const boost::regex activeAnonRegex(R"!(active_anon\s*(\d+)\s*)!");
+
+  // Find out what cgroup version (v1 or v2) we have based on the directory it's
+  // mounted.
+  static const char* cgroupV1Path = "/sys/fs/cgroup/memory/memory.stat";
+  static const char* cgroupV2Path = "/sys/fs/cgroup/memory.stat";
+  if (statFile.empty()) {
+    struct stat buffer;
+    if ((stat(cgroupV1Path, &buffer) == 0)) {
+      statFile = cgroupV1Path;
+    } else if ((stat(cgroupV2Path, &buffer) == 0)) {
+      statFile = cgroupV2Path;
+    } else {
+      statFile = "None";
+    }
+    LOG(INFO) << fmt::format("Using memory stat file {}", statFile);
+  }
+
+  if (statFile != "None") {
+    folly::gen::byLine(statFile.c_str()) |
+        [&](folly::StringPiece line) -> void {
+      inactiveAnon = matchLineWithRegex(line, inactiveAnonRegex, match);
+      activeAnon = matchLineWithRegex(line, activeAnonRegex, match);
+    };
+
+    // Unit is in bytes.
+    return inactiveAnon + activeAnon;
+  }
+
+  // Default case variables.
+  static const boost::regex memAvailableRegex(
+      R"!(MemAvailable:\s*(\d+)\s*kB)!");
+  static const boost::regex memTotalRegex(R"!(MemTotal:\s*(\d+)\s*kB)!");
+  // Last resort use host machine info.
+  folly::gen::byLine("/proc/meminfo") | [&](folly::StringPiece line) -> void {
+    memAvailable = matchLineWithRegex(line, memAvailableRegex, match) * 1024;
+    memTotal = matchLineWithRegex(line, memTotalRegex, match) * 1024;
+  };
+  // Unit is in bytes.
+  return (memAvailable && memTotal) ? memTotal - memAvailable : 0;
+
+#else
+  return 0;
+#endif
+}
+
+size_t LinuxMemoryChecker::matchLineWithRegex(
+    const folly::StringPiece& line,
+    const boost::regex& regex,
+    boost::cmatch& match) {
```

Comment:
The function takes in a output parameter `match` to be more "generic". However, `match` is never used outside. Also the function itself is coded such that it has information about the regex itselft and the result of the match. Namely, that the match contains a numeric value that is converted to size_t.
That means it is specialized and not generic. So I would remove the `match` parameter and rename the function from the more generic name.
Maybe something like `extractNumericConfigValueWithRegex` or something similar.

Reply:
+1

Reply:
Removing match parameter and renaming to suggested name